File: array/15.py
from typing import List
# 超时了
# class Solution:
#     def threeSum(self, nums: List[int]) -> List[List[int]]:
#         if len(nums) < 3 : return []
#         ret = []
#         for i in range(len(nums)):
#             ret.extend(self.twoSum(nums, i))
#         temp = list(set(ret))
#         return [list(tup) for tup in temp]

#     def twoSum(self, nums, idx_target):
#         ret = []
#         for i in range(len(nums)):
#             for j in range(len(nums)):
#                 if i != j and i != idx_target and j != idx_target and nums[i] + nums[j] + nums[idx_target] == 0:
#                     ret.append(tuple(sorted([nums[i], nums[j], nums[idx_target]])))
#         return ret

# 定一议二，确定最小的
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        ret = []
        nums = sorted(nums)
        i = 0
        while i <=  len(nums) - 3:
            if nums[i] > 0:
                break
            target = nums[i]
            low = i + 1
            high =  len(nums) - 1
            while low < high:
                if nums[low] + nums[high] + target < 0:
                    low += 1
                    while low < high and nums[low] == nums[low - 1]:
                        low += 1
                elif nums[low] + nums[high] + target == 0:
                    ret.append([target, nums[low], nums[high]])
                    low += 1
                    high -= 1
                    while low < high and nums[low] == nums[low - 1]:
                        low += 1
                    while low < high and nums[high] == nums[high + 1]:
                        high -= 1
                else:
                    high -= 1
                    while low < high and nums[high] == nums[high + 1]:
                        high -= 1
            i += 1
            while i <= len(nums) - 3 and nums[i] == nums[i - 1]:
                i += 1
        return ret


def main():
    nums = [0, 0, 0, 0]
    print(Solution().threeSum(nums))

if __name__ == "__main__":
    main()

File: leetcode-cpp/15.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if (nums.size() <= 2)
            return vector<vector<int>>();
        // 定一议二，确定最小值
        sort(nums.begin(), nums.end());

//        for (int i= 0; i < nums.size(); i++) {
//            cout << nums[i] << "\t";
//        }
//        cout <<endl;

        int base = 0, i, j, temp;
        vector<vector<int>> ret;
        for (int base = 0; base < nums.size() - 2; base++) {
            if (nums[base] >0 || ( base > 0 && nums[base] == nums[base-1]))
                continue;
            i = base + 1, j = nums.size() - 1;
            while (i < j){
                bool l = false, r = false;
                temp = nums[base] + nums[i] + nums[j];
                if (temp == 0) {
                    ret.push_back(vector<int>({nums[base], nums[i], nums[j]}));
                    l = true;
                    r = true;
                } else if (temp < 0){
                    l = true;
                } else {
                    r = true;
                }
                if (l) {
                    i++;
                    while (i < j && nums[i] == nums[i-1])
                        i++;
                }
                if (r) {
                    j--;
                    while (i < j && nums[j] == nums[j+1])
                        j--;
                }
            }
        }
        return ret;
    }
};
int main(){
    Solution s;
    vector<int> temp({1, -1, -1, 0});
    vector<vector<int>> ret = s.threeSum(temp);
    for (int i = 0; i < ret.size(); i++){
        for (int j = 0; j < ret[i].size(); j++){
            cout << ret[i][j] << "\t";
        }
        cout <<endl;
    }
    return 0;
}
